test/Microsoft.AspNetCore.Blazor.Build.Test/RenderingRazorIntegrationTest.cs

```diff
@@ -43,15 +43,14 @@ public void SupportsCSharpExpressions()
             // Assert
             var frames = GetRenderTree(component);
             Assert.Collection(frames,
-                frame => AssertFrame.Whitespace(frame, 0),
-                frame => AssertFrame.Text(frame, "Hello", 1),
-                frame => AssertFrame.Whitespace(frame, 2),
-                frame => AssertFrame.Whitespace(frame, 3), // @((object)null)
-                frame => AssertFrame.Whitespace(frame, 4),
-                frame => AssertFrame.Text(frame, "123", 5),
-                frame => AssertFrame.Whitespace(frame, 6),
-                frame => AssertFrame.Text(frame, new object().ToString(), 7),
-                frame => AssertFrame.Whitespace(frame, 8));
+                frame => AssertFrame.Text(frame, "Hello", 0),
+                frame => AssertFrame.Whitespace(frame, 1),
+                frame => AssertFrame.Whitespace(frame, 2), // @((object)null)
+                frame => AssertFrame.Whitespace(frame, 3),
+                frame => AssertFrame.Text(frame, "123", 4),
+                frame => AssertFrame.Whitespace(frame, 5),
+                frame => AssertFrame.Text(frame, new object().ToString(), 6),
+                frame => AssertFrame.Whitespace(frame, 7));
```

Comment:
Removing the leading whitespace frame is great. Pretty soon we'll probably want to remove trailing ones too (because I have an important runtime optimisation that is enabled only in the common case where a component renders exactly one top-level node, and trailing whitespace in source docs tends to interfere with that).

If you know an easy way to strip trailing whitespace frames in the Razor compiler please let me know. I was going to implement a compiler pass to do it.

Reply:
Compiler pass is the right way to do it.